### PR TITLE
Treat `k#ident` keywords as valid tokens

### PR DESCRIPTION
**Pull request by fee1-dead**
We treat `k#ident` as a valid token, since it could be useful in implementing experimental features without introducing too much frontend changes (for example, with deref patterns). However, this PR as is would also make the following code compile:

```rust
macro_rules! a {
    (k#test) => {}
}

fn main() {
    a!(k#test);
}
```

We'd probably want to prevent that from happening, so should we feature gate this specific token?

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs [hunk 1]
```diff
@@ -299,6 +299,10 @@ pub enum TokenKind {
     /// It's recommended to use `Token::(ident,uninterpolate,uninterpolated_span)` to
     /// treat regular and interpolated identifiers in the same way.
     Ident(Symbol, /* is_raw */ bool),
+
+    /// A `k#ident` keyword
+    Keyword(Symbol),
```
Comment (Nilstrieb) [inline compiler/rustc_ast/src/token.rs:304]: Actual keywords are Ident, so it seems confusing for a "forced keyword" to be special here.

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
> We'd probably want to prevent that from happening, so should we feature gate this specific token?

Yes, I think we should feature gate `k#` and stabilize it separately and only if some stable feature would make use of it.

**Comment by fee1-dead** [comment]
just converted `token::Ident` to use `IdentKind` - feature gating and tests to follow.

**Comment by Veykril** [comment]
The CI failure is due to https://github.com/rust-lang/rust/pull/118861

**Comment by fee1-dead** [comment]
closed due to design concerns, there is an alternative - using `builtin #` (with a space) and this isn't really useful.

**Comment by fee1-dead** [comment]
No energy to revive this PR, so closing

**Pull request closed without merging**